Reqnroll/Formatters/PubSub/CucumberMessagePublisher.cs

```diff
@@ -0,0 +1,461 @@
+
+        switch (hookBindingStartedEvent.HookBinding.HookType)
+        {
+            case Bindings.HookType.BeforeTestRun:
+            case Bindings.HookType.AfterTestRun:
+            case Bindings.HookType.BeforeFeature:
+            case Bindings.HookType.AfterFeature:
+                var hookRunStartedId = SharedIdGenerator.GetNewId();
+                var hookId = _stepDefinitionIdByMethodBinding[hookBindingStartedEvent.HookBinding];
+                var hookTracker = new TestRunHookExecutionTracker(hookRunStartedId, hookId, _testRunStartedId, _messageFactory);
+                _testRunHookTrackers.TryAdd(hookBindingStartedEvent.HookBinding, hookTracker);
+
+                hookTracker.ProcessEvent(hookBindingStartedEvent);
+                _messages.AddRange(((IGenerateMessage)hookTracker).GenerateFrom(hookBindingStartedEvent));
+                return Task.CompletedTask;
+
+            default:
+                var featureTracker = GetFeatureTracker(hookBindingStartedEvent);
+                featureTracker?.ProcessEvent(hookBindingStartedEvent);
+                break;
+        }
+
+        return Task.CompletedTask;
+    }
+
+    private Task HookBindingFinishedEventHandler(HookBindingFinishedEvent hookBindingFinishedEvent)
+    {
+        if (!_enabled)
+            return Task.CompletedTask;
+
+        switch (hookBindingFinishedEvent.HookBinding.HookType)
+        {
+            case Bindings.HookType.BeforeTestRun:
+            case Bindings.HookType.AfterTestRun:
+            case Bindings.HookType.BeforeFeature:
+            case Bindings.HookType.AfterFeature:
+                if (!_testRunHookTrackers.TryGetValue(hookBindingFinishedEvent.HookBinding, out var hookTracker)) // should not happen
+                    return Task.CompletedTask;
+                hookTracker.ProcessEvent(hookBindingFinishedEvent);
+
+                _messages.AddRange(((IGenerateMessage)hookTracker).GenerateFrom(hookBindingFinishedEvent));
+                return Task.CompletedTask;
+
+            default:
+                var featureTracker = GetFeatureTracker(hookBindingFinishedEvent);
+                featureTracker?.ProcessEvent(hookBindingFinishedEvent);
+                break;
+        }
+
+        return Task.CompletedTask;
+    }
+
+    private Task AttachmentAddedEventHandler(AttachmentAddedEvent attachmentAddedEvent)
+    {
+        if (!_enabled)
+            return Task.CompletedTask;
+
+        var featureInfo = attachmentAddedEvent.FeatureInfo;
+        if (featureInfo == null || string.IsNullOrEmpty(attachmentAddedEvent.ScenarioInfo?.Title))
+        {
+            // This is a TestRun-level attachment (not tied to any feature or scenario)
+            var attachmentIssuedByHookId = ActiveTestRunHook == null ? "" : ActiveTestRunHook.HookStartedId;
+            Debug.Assert(!string.IsNullOrEmpty(attachmentIssuedByHookId), "AttachmentAddedEvent without a FeatureInfo or ScenarioInfo should be issued by a TestRun Hook.");
```

Comment:
`Debug.Assert` fails very badly, so I would try to avoid it. Maybe in this case we should just emit a log entry and ignore the event if it cannot be processed without a hook id.
Same comment applies to line 434 in the `OutputAddedEventHandler` method.

Reply:
Completed (but noted that `Debug.Assert` is present elsewhere throughout the code-base).